reasoner/Reasoner.java

```diff
@@ -158,42 +187,70 @@ public FunctionalIterator<ConceptMap> executeTraversal(Disjunction disjunction,
-                    .map(conceptMgr::conceptMap).filter(answer -> !iterate(conjunction.negations()).flatMap(
-                            negation -> iterator(negation.disjunction(), answer)).hasNext()
-                    ).map(answer -> answer.filter(filter)).distinct();
+                    .map(conceptMgr::conceptMap).filter(answer -> !isNegated(answer, conjunction.negations()))
+                    .map(answer -> answer.filter(filter)).distinct();
         }
     }
 
+    private boolean isNegated(ConceptMap answer, List<Negation> negations) {
+        return iterate(negations).flatMap(n -> iterator(n.disjunction(), answer)).first().isPresent();
+    }
+
     private FunctionalIterator<ConceptMap> iterator(Disjunction disjunction, ConceptMap bounds) {
         return iterate(disjunction.conjunctions()).flatMap(c -> iterator(c, bounds));
     }
 
     private FunctionalIterator<ConceptMap> iterator(Conjunction conjunction, ConceptMap bounds) {
-        return iterator(bound(conjunction, bounds), set());
+        return iterator(bound(conjunction, bounds), Filter.create(list()));
     }
 
-    private FunctionalIterator<ConceptMap> iterator(Conjunction conjunction,
-                                                    Set<Identifier.Variable.Retrievable> filter) {
+    private FunctionalIterator<ConceptMap> iterator(Conjunction conjunction, Filter filter) {
         if (!conjunction.isCoherent()) return Iterators.empty();
-        if (conjunction.negations().isEmpty()) {
-            return traversalEng.iterator(conjunction.traversal(filter)).map(conceptMgr::conceptMap);
-        } else {
-            return traversalEng.iterator(conjunction.traversal()).map(conceptMgr::conceptMap).filter(
-                    ans -> !iterate(conjunction.negations()).flatMap(n -> iterator(n.disjunction(), ans)).hasNext()
-            ).map(conceptMap -> conceptMap.filter(filter)).distinct();
+        FunctionalIterator<ConceptMap> answers = traversalEng.iterator(conjunction.traversal(filter)).map(conceptMgr::conceptMap);
+        if (conjunction.negations().isEmpty()) return answers;
+        else {
+            return traversalEng.iterator(conjunction.traversal()).map(conceptMgr::conceptMap)
+                    .filter(ans -> !isNegated(ans, conjunction.negations()))
+                    .map(conceptMap -> conceptMap.filter(filter)).distinct();
+        }
+    }
+
+    private SortedIterator<ConceptMap.Ordered, SortedIterator.Order.Asc> iteratorSorted(Conjunction conjunction,
+                                                                                        Filter filter, Sorting sorting) {
+        ConceptMap.Ordered.Comparator comparator = ConceptMap.Ordered.Comparator.create(sorting);
+        SortedIterator<ConceptMap.Ordered, SortedIterator.Order.Asc> answers = traversalEng.iterator(conjunction.traversal(filter, sorting))
+                .mapSorted(vertexMap -> conceptMgr.conceptMapOrdered(vertexMap, comparator), ASC);
```

Comment:
execute a traversal, knowing that the answers will respect the sorting required, then map it to the ordered iterator of comparable concept maps

Reply:
this is a compromise between creating a parallel API that returns a sorted iterator of comparable `VertexMap` on the traversal engine (more type safe) and simplicity...